### PR TITLE
Add role management API

### PR DESCRIPTION
**Pull request by ashensw**
Resolves wso2/product-is#8591

Depends on https://github.com/wso2/charon/pull/287, https://github.com/wso2/carbon-identity-framework/pull/3021, https://github.com/wso2/carbon-identity-framework/pull/3022

### CODE REVIEW

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+    @Override
+    public Role createRole(Role role) throws CharonException, ConflictException, BadRequestException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Creating role: " + role.getDisplayName());
+        }
+        try {
+            // Check if the role already exists.
+            if (roleManagementService.isExistingRole(role.getId(), tenantDomain)) {
+                String error = "Role with name: " + role.getDisplayName() + " already exists in the system.";
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java:83]: shall we log other context info such as tenant domain as well?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+            Role createdRole = new Role();
+            createdRole.setId(roleBasicInfo.getId());
+            String locationURI = SCIMCommonUtils.getSCIMRoleURL(roleBasicInfo.getId());
+            createdRole.setLocation(locationURI);
+            createdRole.setDisplayName(roleBasicInfo.getName());
+            createdRole.setSchemas();
+
+            return createdRole;
+        } catch (IdentityRoleManagementException e) {
+            throw new CharonException(
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java:98]: if the exception is from regex validation failure then shouldn't we throw bad request exception here? better to check the error code and send the expected error accordingly

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+                    String.format("Error occurred while adding a new role: %s", role.getDisplayName()), e);
+        }
+    }
+
+    @Override
+    public Role getRole(String roleID, Map<String, Boolean> requiredAttributes)
+            throws BadRequestException, CharonException {
+
+        try {
+            org.wso2.carbon.identity.role.mgt.core.Role role = roleManagementService.getRole(roleID, tenantDomain);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java:108]: let's return not found exception if role is null

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+        } catch (IdentityRoleManagementException e) {
+            throw new CharonException(String.format("Error occurred while getting the role: %s", roleID), e);
+        }
+    }
+
+    @Override
+    public void deleteRole(String roleID) throws CharonException {
+
+        try {
+            roleManagementService.deleteRole(roleID, tenantDomain);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java:148]: validate whether role exists and return not found exception if not.

Comment (ashensw) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java:157]: Fixed and updated the PR.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java (hunk 3 above)
Comment (ashensw) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java:114]: Fixed and updated the PR.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java (hunk 2 above)
Comment (ashensw) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java:105]: Fixed and updated the PR.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java (hunk 1 above)
Comment (ashensw) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMRoleManager.java]: Fixed and updated the PR.

File: pom.xml [hunk 5]
```diff
@@ -295,11 +301,11 @@
         <inbound.auth.oauth.version>6.2.0</inbound.auth.oauth.version>
         <commons-collections.version>3.2.0.wso2v1</commons-collections.version>
         <carbon.kernel.version>4.6.0</carbon.kernel.version>
-        <identity.framework.version>5.17.101</identity.framework.version>
+        <identity.framework.version>5.18.10-SNAPSHOT</identity.framework.version>
```
Comment (emswbandara) [inline pom.xml:304]: let;s update to a released version

File: pom.xml [hunk 6]
```diff
@@ -295,11 +301,11 @@
...
         <junit.version>4.11</junit.version>
         <commons.lang.version>20030203.000129</commons.lang.version>
         <identity.governance.version>1.3.11</identity.governance.version>
-        <charon.version>3.3.17</charon.version>
+        <charon.version>3.3.18-SNAPSHOT</charon.version>
```
Comment (emswbandara) [inline pom.xml:308]: let;s update to a released version

**Pull request merged**